src/test/versionbits_tests.cpp

```diff
@@ -0,0 +1,454 @@
+
+#define VERSIONBITS_UNIT_TEST
+
+#include "chain.h"
+#include "consensus/versionbits.h"
+#include "primitives/block.h"
+#include "test/test_bitcoin.h"
+
+#include <bitset>
+#include <boost/test/unit_test.hpp>
+#include <iomanip>
+#include <iostream>
+#include <sstream>
+#include <stdlib.h>
+#include <string>
+#include <time.h>
+
+using namespace Consensus::VersionBits;
+using namespace std;
+
+const int NBITS = MAX_BIT + 1 - MIN_BIT;
+
+class BitCounter
+{
+public:
+    BitCounter(const Deployments& deployments) :
+        m_deployments(deployments)
+    {
+        Clear();
+    }
+
+    void Clear()
+    {
+        for (int i = 0; i < NBITS; i++)
+            m_bitCounts[i] = 0;
+    }
+
+    void CountBits(int nVersion, uint32_t nTime = 0)
+    {
+        for (int i = MIN_BIT; i <= MAX_BIT; i++)
+        {
+            if (((nVersion >> i) & 0x1) && (m_deployments.GetAssignedRule(i, nTime) != NO_RULE))
+                m_bitCounts[i - MIN_BIT]++;
+        }
+    }
+
+    int GetCountForBit(int bit) const
+    {
+        return m_bitCounts[bit - MIN_BIT];
+    }
+
+    string ToString()
+    {
+        stringstream ss;
+        for (int i = 0; i < NBITS; i++)
+        {
+            if (m_bitCounts[i + MIN_BIT])
+               ss << setw(4) << right << i + MIN_BIT << ": " << setw(5) << right << m_bitCounts[i + MIN_BIT] << endl;
+        }
+
+        return ss.str();
+    }
+
+private:
+    int m_bitCounts[NBITS];
+    const Deployments& m_deployments;
+};
+
+class VersionGenerator
+{
+public:
+    VersionGenerator() { ClearBitProbabilities(); }
+
+    void ClearBitProbabilities()
+    {
+        for (int i = 0; i < NBITS; i++)
+            m_bitProbabilities[i] = 0;
+    }
+
+    void SetBitProbability(int bit, int probability)
+    {
+        m_bitProbabilities[bit - MIN_BIT] = probability;
+    }
+
+    int Generate() const
+    {
+        int nVersion = VERSION_HIGH_BITS;
+        for (int i = 0; i < NBITS; i++)
+        {
+            if ((rand() % ACTIVATION_INTERVAL) <= (m_bitProbabilities[i] - 1))
+                nVersion |= 0x1 << (MIN_BIT + i);
+        }
+
+        return nVersion;
+    }
+    
+private:
+    int m_bitProbabilities[NBITS]; // in units of 1/ACTIVATION_INTERVAL
+};
+
+
+std::string RuleStateToString(State state)
+{
+    switch (state)
+    {
+    case UNDEFINED:
+        return "UNDEFINED";
+
+    case DEFINED:
+        return "DEFINED";
+
+    case LOCKED_IN:
+        return "LOCKED_IN";
+
+    case ACTIVE:
+        return "ACTIVE";
+
+    case FAILED:
+        return "FAILED";
+
+    default:
+        return "N/A";
+    }
+}
+void StateChanged(const CBlockIndex* pblockIndex, const SoftFork* psoftFork, State prevState, State newState, int bitCount)
+{
+    int bit = psoftFork->GetBit();
+    bool isBitSet = (pblockIndex->pprev->nVersion >> bit) & 0x1;
+
+    cout << "STATE CHANGED - height: " << pblockIndex->nHeight << " median time: " << pblockIndex->GetMedianTimePast()
```

Comment:
I like the extra information, but you may look into [BOOST_TEST_MESSAGE()](http://www.boost.org/doc/libs/1_55_0/libs/test/doc/html/utf/user-guide/test-output/BOOST_TEST_MESSAGE.html).